### PR TITLE
Add an API to manage projects, and use it pull their Docker images.

### PR DESCRIPTION
**Pull request by jankeromnes**

### CODE REVIEW

**Comment by jankeromnes** [comment]
Potential problem with this implementation: Pull API is public, and can be spammed by anyone. Maybe we should rate limite?

File: api/projects-api.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+// API sub-resource to manage a single software project.
+const projectAPI = projectsAPI.api('/:project');
+
+projectAPI.post('pull', {
+  title: 'Pull a project',
+  description: 'Trigger a Docker image pull for a given software project.',
+
+  handler: (request, response) => {
+    const { project: projectId } = request.query;
```
Comment (nt1m) [inline api/projects-api.js]: nit: Probably more simple to do:
```
const projectId = request.query.project;
```

File: api/projects-api.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    const projects = db.get('projects');
+    if (!projects[projectId]) {
+      response.statusCode = 404;
```
Comment (nt1m) [inline api/projects-api.js]: nit: `response.statusCode = 404; // Not found`

File: api/projects-api.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      response.json({ error: 'Project not found' }, null, 2);
+      return;
+    }
+
+    machines.pull(projectId, error => {
+      if (error) {
+        log('[fail] pulling project', projectId, error);
+        response.statusCode = 500;
```
Comment (nt1m) [inline api/projects-api.js]: nit: `response.statusCode = 500; // Internal Error`

File: api/projects-api.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+});
+
+// API sub-resource to manage a single software project.
+const projectAPI = projectsAPI.api('/:project');
+
+projectAPI.post('pull', {
+  title: 'Pull a project',
+  description: 'Trigger a Docker image pull for a given software project.',
+
+  handler: (request, response) => {
```
Comment (nt1m) [inline api/projects-api.js]: This API shouldn't be public.

Maybe limit this to logged in users only, or even just admins (`users.isAdmin(user)`)

Comment (jankeromnes) [inline api/projects-api.js]: The thing is that I'd also like to support Docker Hub web hooks. Maybe there is a way to authenticate that as well, eg with a shared secret.

Comment (nt1m) [inline api/projects-api.js]: So, maybe using authentication would be a good choice. Or maybe just rate limit the API.

Comment (jankeromnes) [inline api/projects-api.js]: Hm, I'd love to make this API public somehow, or at least very easily accessible to Docker Hub web hooks, but let's do that in another pull request. I'll make this handler admin-only to begin with.

**Comment by jankeromnes** [comment]
I added a `docker.inspectImage()` method in order to verify the creation date of the image.
- If the image is **more recent** than our current production image, we tag it with `:janitor-production` (and set `project.data.updated` to the creation date, not the pull date)
- If the image is **as old as, or older than** our current production image (e.g. we already pulled this image, or we have an incremental update image that is more recent), then we still keep it, but we don't tag it with `:janitor-production` (and we don't change `project.data.updated`).

File: api/projects-api.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    const { user } = request;
+    // FIXME: Make this API handler accessible to Docker Hub web hooks, so that
+    // it's easy to automatically deploy new project images.
+    if (!users.isAdmin(user)) {
+      response.statusCode = 403; // Forbidden
+      response.json({ error: 'Unauthorized' }, null, 2);
+      return;
+    }
+
+    const projectId = request.query.project;
```
Comment (nt1m) [inline api/projects-api.js]: nit: above, use `const { user, query } = request;` and here: `const projectId = query.project;`

File: api/projects-api.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    machines.pull(projectId, error => {
+      if (error) {
+        log('[fail] pulling project', projectId, error);
+        response.statusCode = 500; // Internal Server Error
+        response.json({ error: 'Could not pull project' }, null, 2);
+        return;
+      }
+
+      response.json({}, null, 2);
```
Comment (nt1m) [inline api/projects-api.js]: Not sure what's the point of this response. Why not just response.end() ?

**Comment by jankeromnes** [comment]
Oh, so clicking on "Update branch" creates a merge commit... 

giphy

**Pull request merged**